sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java

```diff
@@ -34,21 +38,66 @@
  */
 @ThreadSafe
 public final class DirectStreamObserver<T> implements StreamObserver<T> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DirectStreamObserver.class);
+  private static final int DEFAULT_MAX_MESSAGES_BEFORE_CHECK = 100;
+
   private final Phaser phaser;
   private final CallStreamObserver<T> outboundObserver;
+  private final int maxMessagesBeforeCheck;
+
+  private int numberOfMessagesBeforeReadyCheck;
 
   public DirectStreamObserver(
       Phaser phaser,
       CallStreamObserver<T> outboundObserver) {
+    this(phaser, outboundObserver, DEFAULT_MAX_MESSAGES_BEFORE_CHECK);
+  }
+
+  DirectStreamObserver(
+      Phaser phaser,
+      CallStreamObserver<T> outboundObserver,
+      int maxMessagesBeforeCheck) {
     this.phaser = phaser;
     this.outboundObserver = outboundObserver;
+    this.maxMessagesBeforeCheck = maxMessagesBeforeCheck;
   }
 
   @Override
   public void onNext(T value) {
-    int phase = phaser.getPhase();
-    if (!outboundObserver.isReady()) {
-      phaser.awaitAdvance(phase);
+    numberOfMessagesBeforeReadyCheck += 1;
+    if (numberOfMessagesBeforeReadyCheck >= maxMessagesBeforeCheck) {
```

Comment:
This add-check-reset logic seems like it could belong in a single call, or belong to a single object.

Reply:
Separate object would have us just copy forward all this information and will have limited or non-existent re-use.

The logging about gRPC would likely not be relevant anymore.

It could be a separate method but then its just going to read like:
maybeCheckReady();
sync(observer) { observer.onNext(); }
with maybeCheckReady() now being a layer of indirection that a user needs to know about.